torch/csrc/dynamo/guards.cpp

```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+  // and pybind does not accept a unique_ptr reference return type.
+  virtual std::vector<GuardManager*> get_child_managers() {
+    std::vector<GuardManager*> ret;
+    for (const auto& accessor : _accessors) {
+      ret.emplace_back(accessor->get_guard_manager().get());
+    }
+    return ret;
+  }
+
+  // DEBUG function - Returning raw pointers because we can't return unique_ptr
+  // and pybind does not accept a unique_ptr reference return type.
+  std::vector<LeafGuard*> get_leaf_guards() const {
+    std::vector<LeafGuard*> ret;
+    for (const auto& guard : _leaf_guards) {
+      ret.push_back(guard.get());
+    }
+    return ret;
+  }
+
+ protected:
+  // Keeps a count of how many times this guard manager check function returns
+  // False. This is used for sorting optimization.
+  int _fail_count{0};
+
+ private:
+  // Root of the guard manager, this is the used to install the relational
+  // guard resetters.
+  RootGuardManager* _root;
+
+  // Leaf guards are the terminal guards on this object, e.g, type check on a
+  // list. These guards have to be run before any children are run.
+  //
+  // These leaf guards are not shufflable. In almost all cases, these guards
+  // will have an order, e,g., type(x) is int guard and x == 5 guard. We also
+  // expect very few leaf guards per GuardManager node.
+  //
+  // NB: Why are leaf guards shared ptr? This is primarily to enable relational
+  // guards like `tensor X is not tensor Y`. These guards require multiple
+  // values. We handle it by creating one guard object that holds state and this
+  // guard is installed in many guard managers, hence a shared ptr.
+  std::vector<std::shared_ptr<LeafGuard>> _leaf_guards;
+
+  // GuardAccessors nodes to access the child guards. These guards are
+  // shufflable. On a guard failure, they are sorted based on their fail count
+  // to enable fail fast for the next check.
+  std::vector<std::unique_ptr<GuardAccessor>> _accessors;
+};
+
+/**
+ * RootGuardManager is the root of the guard tree. This is primarily
+ * constructed to hold the relational guard pointers so that we can reset the
+ * state of those guards on guard failure. All the other important
+ * implementation is in GuardManager class.
+ */
+
+class RootGuardManager : public GuardManager {
+ public:
+  // This is the root node, set its _root member to nullptr
+  RootGuardManager() : GuardManager(this) {}
+
+  // Adds the relational guard resetter
+  void add_relational_guard_resetter(
+      std::shared_ptr<RelationalGuard> relational_guard) {
+    _relational_guard_resetters.emplace_back(std::move(relational_guard));
+  }
+
+  // Python visible API to check guard function.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  // Python visible API to check_verbose guard function.
+  GuardDebugInfo check_verbose(py::handle value) {
+    return check_verbose_nopybind(value.ptr());
+  }
```

Comment:
weren't these already declared in base class?

Reply:
No. We only want these to for `RootGuardManager`. We declared for `LeafGuard` but thats a separate data structure.